File: whisper/ZeticMLangeWhisper-Android/app/src/main/java/com/zeticai/zeticmlangewhisper_android/ProbabilityUtils.kt
package com.zeticai.zeticmlangewhisper_android

import kotlin.math.exp
import kotlin.random.Random

class ProbabilityUtils {
    companion object {
        fun softmax(logits: FloatArray): FloatArray {
            val maxLogit = logits.maxOrNull() ?: 0f
            val expLogits = FloatArray(logits.size) { i ->
                exp((logits[i] - maxLogit).toDouble()).toFloat()
            }
            val sumExp = expLogits.sum()
            return FloatArray(logits.size) { i -> expLogits[i] / sumExp }
        }

        fun sampleFromDistribution(probs: FloatArray): Int {
            val random = Random.Default.nextFloat()
            var cumSum = 0f
            for (i in probs.indices) {
                cumSum += probs[i]
                if (random < cumSum) {
                    return i
                }
            }
            return probs.size - 1
        }

        fun argmax(array: FloatArray): Int {
            var maxIndex = 0
            var maxValue = array[0]
            for (i in 1 until array.size) {
                if (array[i] > maxValue) {
                    maxIndex = i
                    maxValue = array[i]
                }
            }
            return maxIndex
        }
    }
}

File: whisper/ZeticMLangeWhisper-iOS/ZeticMLangeWhisperiOS/ProbabilityUtils.swift
import Foundation

class ProbabilityUtils {
    static func softmax(_ logits: [Float]) -> [Float] {
        guard let maxLogit = logits.max() else { return [] }

        let expLogits = logits.map { Float(exp(Double($0 - maxLogit))) }
        let sumExp = expLogits.reduce(0, +)

        return expLogits.map { $0 / sumExp }
    }

    static func sampleFromDistribution(_ probs: [Float]) -> Int {
        let random = Float.random(in: 0 ..< 1)
        var cumSum: Float = 0

        for (index, prob) in probs.enumerated() {
            cumSum += prob
            if random < cumSum {
                return index
            }
        }

        return probs.count - 1
    }

    static func argmax(_ array: [Float]) -> Int {
        guard !array.isEmpty else { return 0 }

        var maxIndex = 0
        var maxValue = array[0]

        for (index, value) in array.enumerated() {
            if value > maxValue {
                maxIndex = index
                maxValue = value
            }
        }

        return maxIndex
    }
}
